### PR TITLE
Completed Sign in and Log in

### PR DESCRIPTION
**Pull request by SemanBoon**
**SUMMARY**

**Features Implemented**
The User and Service Provider Signup Page:
1. Takes in basic sign up information like user's business name, full name, email, phone number, and password.
2. Validates the input fields to ensure correct formatting (e.g., email format, password length).
3. Sends the data to the backend endpoint that handles the creation of the account.

Login Page:
1. Checks the entered email and password for correct format and if it exists in the database.
2. Checks the entered email in both user and serviceProvider tables to determine the user type.
3. Verifies the password and returns the user type along with the user data if the credentials are valid.

Select Profile Modal:
1. When a user clicks on the "Sign Up" link on the login page, a modal pops up asking the user to select between User and Service Provider.
2. Based on the selection, the user is redirected to the respective signup form.

Improvements and Fixes
1. User Type Check in Login: I updated the login endpoint to check both user and serviceProvider tables to determine the user type based on the entered email.
2. I corrected my code to make sure the correct user type is returned to ensure after successful login.

Prisma Schema Update:
1. I added userType field to both user and serviceProvider tables in the Prisma schema.
2. I ran Prisma migrations to apply the schema changes to the database.

Form Validation and Error Handling:
1. I added validations for the email and password fields in both signup and login forms.
2. I implemented error handling to display appropriate and detailed messages to users in case of invalid input or failed operations.

**LOOM VIDEO LINK** https://www.loom.com/share/4e94a8fc1c594a6e822e75daca62618e?sid=2bf8ab77-dcc5-431f-81f0-11da5306db82

### CODE REVIEW

File: Backend/passport-config.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+// const passport = require('passport');
```
Comment (Scott-Jia) [inline Backend/passport-config.js:3]: @SemanBoon Could you please clean up the code that has already been commented out, or simply add a comment/todo to explain why you are keeping those? same for some others

Comment (SemanBoon) [reply Backend/passport-config.js:3]: They are all base code I need to work on JWT tokens for my stretch feature, I'll add a comment on it and ensure that I clean everything up before creating my next PR.
Thank you bo.

Comment (Scott-Jia) [reply Backend/passport-config.js:3]: Cool, just add what you're saying as a comment! Thanks

File: Backend/index.js [hunk 2]
```diff
@@ -76,6 +70,26 @@ app.post("/login", async (req, res) => {
...
+    });
+    if (!tokenRecord) {
+      return res.status(403).json({ error: 'Invalid refresh token' });
+    }
+    const isValidRefreshToken = jwt.verify(refreshToken, process.env.REFRESH_TOKEN_SECRET);
+    if (!isValidRefreshToken) {
+      return res.status(403).json({ error: 'Token is not valid' });
+    }
+    const user = { id: tokenRecord.userId};
+    const newAccessToken = jwt.sign(user, process.env.ACCESS_TOKEN_SECRET, { expiresIn: '20m' });
```
Comment (gkeerthana23) [inline Backend/index.js]: @SemanBoon Just curious, why did we pick 20m here?

File: Backend/passport-config.js (hunk 1 above)
Comment (gkeerthana23) [reply Backend/passport-config.js:3]: @SemanBoon Like we discussed offline, I would recommend keeping any future code stored locally and committing only once it's finalized and thoroughly tested out :)

File: frontend/src/Components/LoginForm/LoginForm.jsx [hunk 3]
```diff
@@ -16,68 +17,63 @@ const LoginForm = () => {
...
-        console.log('success')
+        const data = await response.json();
+        localStorage.setItem("token", data.accessToken);
+        localStorage.setItem("refreshToken", data.refreshToken);
+        updateUser({ email: userEmail });
+        navigate("/homepage");
       } else {
-        setResult("failed to login");
-        console.log('failed')
+        console.log("Login failed");
```
Comment (gkeerthana23) [inline frontend/src/Components/LoginForm/LoginForm.jsx]: @SemanBoon Console log statements are mainly used for debugging. We should display an error message to the user (which I believe you already are?) and if we want o log errors for data analysis later, we should set up an error table and log errors there. We don't have to do this for the project, but let's clean up any console log statements from the committed code.

File: frontend/src/Components/SignupForm/SignupForm.jsx [hunk 4]
```diff
@@ -26,95 +39,136 @@ const SignupForm = () => {
     setUserName(e.target.value);
   };
 
+  const handleSignup = async (e) => {
+    e.preventDefault();
+
+    const emailPattern = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
```
Comment (gkeerthana23) [inline frontend/src/Components/SignupForm/SignupForm.jsx:47]: @SemanBoon Just trying to understand this a bit better - could you share a bit more on the pattern that we're validating here?

File: frontend/src/Components/SignupForm/SignupForm.jsx [hunk 5]
```diff
@@ -26,95 +39,136 @@ const SignupForm = () => {
...
-      },
-      body: JSON.stringify({
-        name: userName,
-        email: userEmail,
-        phoneNumber: parseInt(userPhone),
-        password: userPassword,
-      }),
-    })
-    .then((response) => {
+    const formattedName = capitalizedName(userName);
```
Comment (gkeerthana23) [inline frontend/src/Components/SignupForm/SignupForm.jsx:67]: @SemanBoon Again just trying to understand - why are we capitalizing name and email?

File: frontend/src/Components/SignupForm/SignupForm.jsx [hunk 6]
```diff
@@ -26,95 +39,136 @@ const SignupForm = () => {
     setUserName(e.target.value);
   };
 
+  const handleSignup = async (e) => {
```
Comment (gkeerthana23) [inline frontend/src/Components/SignupForm/SignupForm.jsx:44]: @SemanBoon What happens if a user or service provider tries to sign up with and email that already exists in our database and has an account associated with it?

File: frontend/src/Utils/fetchWithAuth.jsx [hunk 7]
```diff
@@ -1,47 +1,47 @@
...
-    if (!response.ok) {
-        throw new Error('Failed to refresh access token');
-    }
-
-    const data = await response.json();
-    return data.accessToken;
-};
-
-export default fetchWithAuth;
+// const fetchWithAuth = async (url, options = {}) => {
```
Comment (gkeerthana23) [inline frontend/src/Utils/fetchWithAuth.jsx:3]: @SemanBoon Let's clean this up if we haven't already :)

File: Backend/index.js [hunk 8]
```diff
@@ -63,36 +78,65 @@ app.post("/service-provider-signup", async (req, res) => {
...
+    } else {
+      return res.status(401).json({ error: "Invalid credentials" });
     }
   } catch (e) {
     console.log(e.message)
     res.status(500).json({ error: e.message });
   }
 });
+
+// app.post("/login", async (req, res) => {
```
Comment (gkeerthana23) [inline Backend/index.js]: @SemanBoon Let's clean this up if we haven't already! Let's also get rid of the console log statements throughout .

File: Backend/prisma/schema.prisma [hunk 9]
```diff
@@ -27,10 +27,10 @@ model ServiceProvider {
   id       String   @id @default(cuid())
   email String @unique
   businessName     String
-  businessAddress    String
+  businessAddress    String?
```
Comment (gkeerthana23) [inline Backend/prisma/schema.prisma]: @SemanBoon Why did we make the business address an optional field? Don't we need it to be a required field?

File: Backend/prisma/schema.prisma [hunk 10]
```diff
@@ -27,10 +27,10 @@ model ServiceProvider {
...
   phoneNumber   Int
   hashedPassword String
-  services String
-  profilePhoto String
-  coverPhoto String
+  services String?
```
Comment (gkeerthana23) [inline Backend/prisma/schema.prisma]: @SemanBoon Same with services

File: frontend/src/App.jsx [hunk 11]
```diff
@@ -27,13 +29,13 @@ function App() {
   return (
     <div className="app">
       <UserContext.Provider value={{ user, updateUser }}>
-        <Router>
+      <Router>
```
Comment (gkeerthana23) [inline frontend/src/App.jsx:35]: @SemanBoon Glad we fixed the Router issue!

File: frontend/src/Components/AddressLookupComponent/AddressLookupComponent.jsx [hunk 12]
```diff
@@ -1,44 +1 @@
-// import React, { useState, useEffect } from "react";
```
Comment (gkeerthana23) [inline frontend/src/Components/AddressLookupComponent/AddressLookupComponent.jsx]: @SemanBoon Could you share why we're deleting all the code in this file? Also, reminder to have commit messages reflect the changes that are happening within it :)

**Review by gkeerthana23 — APPROVED** [review]
Merging PR to unblock. @SemanBoon Feel free to take a look at and address comments in follow up commits!

**Pull request merged**